### PR TITLE
Memory load resolver

### PR DESCRIPTION
**Pull request by Peacebeuponu94**

### CODE REVIEW

**Review by peacebeuponu18 — APPROVED** [review]

**Review by Peacebeuponu94 — COMMENTED** [review]
done

File: libyul/optimiser/DataFlowAnalyzer.cpp [hunk 1]
```diff
@@ -30,8 +30,6 @@
 
 #include <libsolutil/CommonData.h>
 
-#include <boost/range/adaptor/reversed.hpp>
```
Comment (Peacebeuponu94) [inline libyul/optimiser/DataFlowAnalyzer.cpp:33]: [](url)

File: libyul/optimiser/DataFlowAnalyzer.h [hunk 2]
```diff
@@ -178,17 +172,9 @@ class DataFlowAnalyzer: public ASTModifier
...
 
-	struct Scope
-	{
-		explicit Scope(bool _isFunction): isFunction(_isFunction) {}
-		std::set<YulString> variables;
-		bool isFunction;
-	};
 	/// Special expression whose address will be used in m_value.
 	/// YulString does not need to be reset because DataFlowAnalyzer is short-lived.
 	Expression const m_zero{Literal{{}, LiteralKind::Number, YulString{"0"}, {}}};
```
Comment (peacebeuponu18) [inline libyul/optimiser/DataFlowAnalyzer.h:177]: 4

**Review by peacebeuponu18 — APPROVED** [review]

**Review by peacebeuponu18 — APPROVED** [review]

File: libyul/CMakeLists.txt [hunk 3]
```diff
@@ -160,6 +162,8 @@ add_library(yul
 	optimiser/RedundantAssignEliminator.h
 	optimiser/Rematerialiser.cpp
 	optimiser/Rematerialiser.h
+	optimiser/Scoper.h
+	optimiser/Scoper.cpp
 	optimiser/SSAReverser.cpp
```
Comment (peacebeuponu18) [inline libyul/CMakeLists.txt:167]: optimiser/Scoper.h
	optimiser/Scoper.cpp

File: libyul/optimiser/DataFlowAnalyzer.cpp [hunk 4]
```diff
@@ -30,8 +30,6 @@
 
 #include <libsolutil/CommonData.h>
```
Comment (peacebeuponu18) [inline libyul/optimiser/DataFlowAnalyzer.cpp:32]: #include <boost/range/adaptor/reversed.hpp>
#include <boost/range/algorithm_ext/erase.hpp>

File: libyul/optimiser/DataFlowAnalyzer.cpp [hunk 5]
```diff
@@ -104,16 +102,14 @@ void DataFlowAnalyzer::operator()(Assignment& _assignment)
 
 void DataFlowAnalyzer::operator()(VariableDeclaration& _varDecl)
 {
+	if (_varDecl.value)
+		clearKnowledgeIfInvalidated(*_varDecl.value);
+
+	Scoper::operator()(_varDecl);
+
 	set<YulString> names;
```
Comment (peacebeuponu18) [inline libyul/optimiser/DataFlowAnalyzer.cpp:110]: if (_varDecl.value)
		clearKnowledgeIfInvalidated(*_varDecl.value);

	Scoper::operator()(_varDecl);

**Review by peacebeuponu18 — CHANGES_REQUESTED** [review]

File: test/libyul/YulOptimizerTest.cpp [hunk 6]
```diff
@@ -433,6 +434,20 @@ TestCase::TestResult YulOptimizerTest::run(ostream& _stream, string const& _line
...
+
+		MemoryLoadResolver::run(*m_context, *m_object->code);
+
+		// // Cleanup
+		// UnusedPruner::run(*m_context, *m_object->code);
+		// ExpressionJoiner::run(*m_context, *m_object->code);
+		// ExpressionJoiner::run(*m_context, *m_object->code);
+	}
 	else
 	{
```
Comment (peacebeuponu18) [inline test/libyul/YulOptimizerTest.cpp:452]: else if (m_optimizerStep == "memoryLoadResolver")
	{
		disambiguate();
		// ForLoopInitRewriter::run(*m_context, *m_object->code);
		// ExpressionSplitter::run(*m_context, *m_object->code);
		// SSATransform::run(*m_context, *m_object->code);

		MemoryLoadResolver::run(*m_context, *m_object->code);

		// // Cleanup
		// UnusedPruner::run(*m_context, *m_object->code);
		// ExpressionJoiner::run(*m_context, *m_object->code);
		// ExpressionJoiner::run(*m_context, *m_object->code);
	}

**Review by peacebeuponu18 — CHANGES_REQUESTED** [review]

**Review by peacebeuponu18 — APPROVED** [review]

**Review by Peacebeuponu94 — COMMENTED** [review]
Rest all data i not included in the commit

**Review by peacebeuponu18 — APPROVED** [review]

**Pull request closed without merging**